File: src/cxx/VulkanLib/Device/Buffer/PushConstant.cpp
#include "PushConstant.hpp"

PushConstant::PushConstant(size_t size, vk::PipelineLayout pipelineLayout) : size(size), layout(pipelineLayout){
    data = malloc(size);
}

void PushConstant::bind(vk::CommandBuffer cmd, vk::ShaderStageFlags bindPoint) {
    cmd.pushConstants(layout, bindPoint, 0, (uint32_t)size, data);
}

void PushConstant::writeData(void *pData, size_t writeSize) {
    if(writeSize>size){
        throw std::runtime_error("Error: you cannot write pData that size exceeds the push constant size");
    }
    memcpy(PushConstant::data, pData, writeSize);
}

void PushConstant::destroy() {
    destroyed = true;
    free(data);
}

File: src/device/buffer/push_constant.rs
use ash::vk;

pub struct VlPushConstant {
    data: Vec<u8>,
    layout: vk::PipelineLayout,
}

impl VlPushConstant {
    pub fn new(size: usize, layout: vk::PipelineLayout) -> Self {
        Self {
            data: vec![0u8; size],
            layout,
        }
    }

    pub fn bind(
        &self,
        device: &ash::Device,
        cmd: vk::CommandBuffer,
        stage_flags: vk::ShaderStageFlags,
    ) {
        unsafe {
            device.cmd_push_constants(
                cmd,
                self.layout,
                stage_flags,
                0,
                &self.data,
            );
        }
    }

    pub fn write_data(&mut self, src: &[u8]) {
        if src.len() > self.data.len() {
            panic!("Push constant write exceeds allocated size");
        }

        self.data[..src.len()].copy_from_slice(src);
    }

    pub fn size(&self) -> usize {
        self.data.len()
    }
}
